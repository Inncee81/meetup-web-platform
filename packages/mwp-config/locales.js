module.exports = [
	'en-US',	
	'en-AU',
	'es',
	'es-ES',
	'de',
	'de-DE',
	'fr',
	'fr-FR',
	'it',
	'it-IT',
	'ja',
	'ja-JP',
	'ko',
	'ko-KR',
	'nl',	
	'nl-NL',
	'pt',
	'pt-BR',
	'pl',
	'pl-PL',
	'ru',
	'ru-RU',
	'th',
	'th-TH',
	'tr',
	'tr-TR',
];
