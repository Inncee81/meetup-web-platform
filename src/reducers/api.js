/**
 * The root level reducer for the app.
 * @module reducer
 **/

import { combineReducers } from 'redux';
import {
	CLICK_TRACK_ACTION,
	CLICK_TRACK_CLEAR_ACTION
} from '../actions/clickActionCreators';

export const DEFAULT_APP_STATE = { isFetching: false };

/**
 * The primary reducer for data provided by the API
 * `state.app` sub-tree
 *
 * @param {Object} state
 * @param {ReduxAction} action
 * @return {Object}
 */
export function app(state=DEFAULT_APP_STATE, action={}) {
	switch (action.type) {
	case 'API_REQUEST':
		if (action.meta.logout) {
			// clear app state during logout
			return { ...DEFAULT_APP_STATE, isFetching: true };
		}
		return { ...state, isFetching: true };
	case 'API_SUCCESS':  // fall though
	case 'CACHE_SUCCESS':  // fall through
	case 'API_ERROR': {
		// each of these actions provides an API response that should go into app
		// state - error responses will contain error info
		delete state.failure;  // if there are any values, the API is not failing
		const { ref, ...response } = action.payload.response;
		return { ...state, [ref]: response };
	}
	case 'API_FAILURE':
		state.failure = action.payload;
		// fall through - fetch is complete
	case 'API_COMPLETE':
		return { ...state, isFetching: false };
	default:
		return state;
	}
}

export const DEFAULT_CLICK_TRACK = { history: [] };
/**
 * @param {Object} data extensible object to store click data {
 *   history: array
 * }
 * @param {Object} action the dispatched action
 * @return {Object} new state
 */
export function clickTracking(state=DEFAULT_CLICK_TRACK, action) {
	if (action.type === CLICK_TRACK_ACTION) {
		const history = [
			...state.history,
			action.payload,
		];
		return {
			...state,
			history,
		};
	}
	if (action.type === CLICK_TRACK_CLEAR_ACTION) {
		return DEFAULT_CLICK_TRACK;
	}
	return state;
}

export function config(state={}, action) {
	switch(action.type) {
	case 'CONFIGURE_API_URL':
		return { ...state, apiUrl: action.payload };
	case 'CONFIGURE_BASE_URL':
		return { ...state, baseUrl: action.payload };
	default:
		return state;
	}
}

/**
 * This reducer manages a list of boolean flags that indicate the 'ready to
 * render' state of the application. It is used exclusively by the server,
 * which triggers actions when initializing a response that should eventually
 * make all flags 'true'
 *
 * The server can then read these flags from state and render when ready
 */
export function preRenderChecklist([apiDataLoaded] = [false], action) {
	return [
		apiDataLoaded || action.type === 'API_COMPLETE',
	];
}

const platformReducers = {
	app,
	clickTracking,
	config,
	preRenderChecklist,
};

/**
 * A function that builds a reducer combining platform-standard reducers and
 * app-specific reducers
 */
export default function makeRootReducer(appReducers={}) {
	Object.keys(appReducers).forEach(reducer => {
		if (reducer in platformReducers) {
			throw new Error(`'${reducer}' is a reserved platform reducer name`);
		}
	});
	return combineReducers({
		...platformReducers,
		...appReducers,
	});
}

