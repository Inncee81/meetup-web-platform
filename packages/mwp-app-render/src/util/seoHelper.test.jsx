import locales from 'mwp-config/locales';
import { MOCK_EVENT } from 'meetup-web-mocks/lib/api';

import {
	generateMetaData,
	generateGeoMetaData,
	generateCanonicalUrl,
	generateCanonicalUrlMetaTags,
	generateEventLdJson,
	generateLocationLdJson,
	generateFeeLdJson,
	generateOrganizationLdJson,
} from './seoHelper';

const convertToLocalTime = jest.fn(() => new Date('2027-01-31T04:20:12.142Z'));

const venue = {
	name: 'some place',
	address_1: 'address 1',
	address_2: 'address 2',
	address_3: 'address 3',
	city: 'city name',
	state: 'NY',
	zip: '00000',
	country: 'usa',
	localized_country_name: 'usa',
};

const fee = {
	amount: 11,
	currency: 'USD',
};

const MOCK_META = {
	title: 'mock title',
	description: 'mock description',
	keywords: 'mock keywords',
	baseUrl: 'https://mock-base-url.com',
	route: '/mock/preview/path',
	appPath: '/mock/app/path',
	imageUrl: 'https://www.meetup.com/mock-default-image.jpg',
	localeCode: 'fr-FR',
};

describe('generateMetaData', () => {
	const metaData = generateMetaData(MOCK_META);

	it('should set the title', () => {
		const titleMeta = metaData.filter(
			obj => obj.content === MOCK_META.title && obj.property === 'og:title'
		);
		expect(titleMeta).toHaveLength(1);
	});

	it('should set the description', () => {
		const descriptionMeta = metaData.filter(
			obj =>
				obj.content === MOCK_META.description &&
				(obj.property === 'og:description' || obj.name === 'description')
		);
		expect(descriptionMeta).toHaveLength(2);
	});

	it('should set the keywords', () => {
		const keywordsMeta = metaData.filter(
			obj => obj.content === MOCK_META.keywords && obj.name === 'keywords'
		);
		expect(keywordsMeta).toHaveLength(1);
	});

	it('should set the app url', () => {
		const appUrlMeta = metaData.filter(obj => obj.property === 'al:web:url');
		expect(appUrlMeta).toHaveLength(1);
	});

	it('should set the appPath', () => {
		const appPathMeta = metaData.filter(obj => obj.content === MOCK_META.appPath);
		expect(appPathMeta).toHaveLength(2);
	});

	it('should set the url as localized url', () => {
		const expectedUrl = generateCanonicalUrl(
			MOCK_META.baseUrl,
			MOCK_META.localeCode,
			MOCK_META.route
		);
		const ogUrlMeta = metaData.filter(
			obj => obj.content === expectedUrl && obj.property === 'og:url'
		);
		const alUrlMeta = metaData.filter(
			obj => obj.content === expectedUrl && obj.property === 'al:web:url'
		);
		expect(ogUrlMeta).toHaveLength(1);
		expect(alUrlMeta).toHaveLength(1);
	});

	describe('open graph meta data', () => {
		it('should set og:site_name to "Meetup"', () => {
			const ogSiteName = obj =>
				obj.content === 'Meetup' && obj.property === 'og:site_name';
			expect(ogSiteName).toHaveLength(1);
		});
		it('should set og:type to "article"', () => {
			const metaData = generateMetaData(MOCK_META);
			const ogType = metaData.filter(
				obj => obj.content === 'article' && obj.property === 'og:type'
			);
			expect(ogType).toHaveLength(1);
		});
		it('should set og:url', () => {
			const urlMeta = metaData.filter(obj => obj.property === 'og:url');
			expect(urlMeta).toHaveLength(1);
		});
		it('should set og:image', () => {
			const ogImage = metaData.filter(
				obj => obj.content === MOCK_META.imageUrl && obj.property === 'og:image'
			);
			expect(ogImage).toHaveLength(1);
		});
	});
	describe('twitter meta data', () => {
		it('should set twitter:image', () => {
			const twitterImage = metaData.filter(
				obj =>
					obj.content === MOCK_META.imageUrl && obj.property === 'twitter:image'
			);
			expect(twitterImage).toHaveLength(1);
		});
	});
});

describe('generateGeoMetaData', () => {
	it('should set position', () => {
		const lat = 'hello world';
		const lon = 'hello world';
		const processedObj = generateGeoMetaData({ lat, lon });
		const filteredObj = processedObj.filter(
			obj => obj.content === `${lat};${lon}` && obj.property === 'geo.position'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should NOT geo position', () => {
		const processedObj = generateGeoMetaData({});
		const filteredObj = processedObj.filter(obj => obj.property === 'geo.position');
		expect(filteredObj).toHaveLength(0);
	});

	it('should set place w/ city', () => {
		const city = 'hello world';
		const processedObj = generateGeoMetaData({ city });
		const filteredObj = processedObj.filter(
			obj => obj.content === city && obj.property === 'geo.placename'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should set place w/ city + state', () => {
		const city = 'hello world';
		const state = 'hello world';
		const processedObj = generateGeoMetaData({ city, state });
		const filteredObj = processedObj.filter(
			obj => obj.content === `${city}, ${state}` && obj.property === 'geo.placename'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should set place w/ city + state + country', () => {
		const city = 'hello world';
		const state = 'hello world';
		const country = 'USA';
		const processedObj = generateGeoMetaData({ city, state, country });
		const filteredObj = processedObj.filter(
			obj =>
				obj.content === `${city}, ${state}, ${country}` &&
				obj.property === 'geo.placename'
		);
		expect(filteredObj).toHaveLength(1);
	});

	it('should set place w/ city + state + country', () => {
		const country = 'USA';
		const processedObj = generateGeoMetaData({ country });
		const filteredObj = processedObj.filter(
			obj => obj.content === country && obj.property === 'geo.region'
		);
		expect(filteredObj).toHaveLength(1);
	});
});

describe('generateCanonicalUrlMetaTags', () => {
	it('should generate locale <link />s for all locales', () => {
		const baseUrl = 'http://www.amaraisthecutest.cat',
			route = '/samsonite-incorporated';
		const canonicalUrls = generateCanonicalUrlMetaTags(baseUrl, '', route);
		// + 2 for default canonical links
		expect(canonicalUrls).toHaveLength(locales.length + 2);
	});

	it('should generate a canonical url with locale if *not* en-US', () => {
		const baseUrl = 'http://www.amaraisthecutest.cat',
			localeCode = 'fr-FR',
			route = '/samsonite-incorporated';
		const processedObj = generateCanonicalUrlMetaTags(baseUrl, localeCode, route);
		const filterArr = processedObj.filter(
			element =>
				element.props.rel === 'canonical' &&
				element.props.href === `${baseUrl}/${localeCode}${route}`
		);
		expect(filterArr).toHaveLength(1);
	});

	it('should generate a canonical url without locale if en-US', () => {
		const baseUrl = 'http://www.amaraisthecutest.cat',
			localeCode = 'en-US',
			route = '/samsonite-incorporated';
		const processedObj = generateCanonicalUrlMetaTags(baseUrl, localeCode, route);
		const filterArr = processedObj.filter(
			element =>
				element.props.rel === 'canonical' &&
				element.props.href === `${baseUrl}${route}`
		);
		expect(filterArr).toHaveLength(1);
	});

	it('should generate a x-default url', () => {
		const baseUrl = 'http://www.amaraisthecutest.cat',
			localeCode = 'fr-FR',
			route = '/samsonite-incorporated';
		const processedObj = generateCanonicalUrlMetaTags(baseUrl, localeCode, route);
		const filterArr = processedObj.filter(
			element =>
				element.props.hrefLang === 'x-default' &&
				element.props.href === `${baseUrl}${route}`
		);
		expect(filterArr).toHaveLength(1);
	});
});

describe('generateFeeLdJson', () => {
	it('should generate a ld+json object for offers', () => {
		const processedObj = generateFeeLdJson(fee);
		expect(processedObj).toMatchSnapshot();
	});
});

describe('generateLocationLdJson', () => {
	it('should generate a ld+json object for location', () => {
		const processedObj = generateLocationLdJson(venue);
		expect(processedObj).toMatchSnapshot();
	});
});

describe('generateEventLdJson', () => {
	it('should generate a ld+json object for event', () => {
		const event = {
			...MOCK_EVENT,
			time: 1501284510,
			venue,
			fee,
		};

		const processedObj = generateEventLdJson(event);

		expect(processedObj).toMatchSnapshot();
	});
});

describe('generateOrganizationLdJson', () => {
	it('should generate json with all the correct fields', () => {
		const baseUrl = 'http://www.craaazyurl.com';
		const localeCode = 'es';
		const route = '/woooo';
		const json = generateOrganizationLdJson(baseUrl, localeCode, route);
		expect(json).toHaveProperty('@type');
		expect(json).toHaveProperty('@context');
		expect(json).toHaveProperty('url');
		expect(json).toHaveProperty('name');
		expect(json).toHaveProperty('logo');
		expect(json).toHaveProperty('sameAs');
		expect(json.sameAs).toHaveLength(5);
	});

	it('should generate localized url and social media links', () => {
		const baseUrl = 'http://www.craaazyurl.com';
		const localeCode = 'fr-FR';
		const route = '/woooo';
		const json = generateOrganizationLdJson(baseUrl, localeCode, route);
		const facebookLink = json.sameAs[0];
		const twitterLink = json.sameAs[1];
		expect(json.url).toBe(`${baseUrl}/${localeCode}${route}`);
		expect(facebookLink).toBe('https://www.facebook.com/MeetupFR/');
		expect(twitterLink).toBe('https://twitter.com/MeetupFR/');
	});
});
