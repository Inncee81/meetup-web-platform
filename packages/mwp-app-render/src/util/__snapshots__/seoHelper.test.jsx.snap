// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateEventLdJson should generate a ld+json object for event 1`] = `
Object {
  "@context": "http://schema.org",
  "@type": "Event",
  "description": "The coolest event in the world during which we will run and dance and sing
	<script>alert(\\"bad time\\")</script>, ðŸ˜Š, &lt;blink&gt;what what&lt;blink&gt; this is getting
	longer than it needs to be why am I still typing omg",
  "location": Object {
    "@type": "Place",
    "address": Object {
      "@type": "PostalAddress",
      "addressCountry": "usa",
      "addressLocality": "city name",
      "addressRegion": "NY",
      "postalCode": "00000",
      "streetAddress": "address 1 address 2 address 3",
    },
    "geo": Object {
      "@type": "GeoCoordinates",
      "latitude": undefined,
      "longitude": undefined,
    },
    "name": "some place",
  },
  "name": "So much fun",
  "offers": Object {
    "@type": "Offer",
    "price": 11,
    "priceCurrency": "USD",
  },
  "startDate": "TODO-START-DATE",
  "url": undefined,
}
`;

exports[`generateFeeLdJson should generate a ld+json object for offers 1`] = `
Object {
  "offers": Object {
    "@type": "Offer",
    "price": 11,
    "priceCurrency": "USD",
  },
}
`;

exports[`generateLocationLdJson should generate a ld+json object for location 1`] = `
Object {
  "location": Object {
    "@type": "Place",
    "address": Object {
      "@type": "PostalAddress",
      "addressCountry": "usa",
      "addressLocality": "city name",
      "addressRegion": "NY",
      "postalCode": "00000",
      "streetAddress": "address 1 address 2 address 3",
    },
    "geo": Object {
      "@type": "GeoCoordinates",
      "latitude": undefined,
      "longitude": undefined,
    },
    "name": "some place",
  },
}
`;
